### PR TITLE
[Security] Remove getPassword() and getSalt() from UserInterface

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

with BC layers from https://github.com/symfony/symfony/pull/40267

### CODE REVIEW

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php [hunk 1]
```diff
@@ -32,30 +32,20 @@ public function __construct(PasswordHasherFactoryInterface $hasherFactory)
 
     public function hashPassword(PasswordAuthenticatedUserInterface $user, string $plainPassword): string
     {
-        $salt = null;
-        if ($user instanceof LegacyPasswordAuthenticatedUserInterface) {
```
Comment (stof) [inline src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: this change will increase merge conflicts for no benefits IMO

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php [hunk 2]
```diff
@@ -51,7 +51,12 @@ public function __construct(UserProviderInterface $userProvider, UserCheckerInte
      */
     protected function checkAuthentication(UserInterface $user, UsernamePasswordToken $token)
     {
+        if ($user instanceof PasswordAuthenticatedUserInterface) {
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: missing `!`

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php [hunk 3]
```diff
@@ -51,7 +51,12 @@ public function __construct(UserProviderInterface $userProvider, UserCheckerInte
...
+            throw new \LogicException(sprintf('Class "%s" must implement "%s" for using password-based authentication.',get_debug_type($user), PasswordAuthenticatedUserInterface::class));
+        }
+
         $currentUser = $token->getUser();
+
         if ($currentUser instanceof UserInterface) {
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: this should also check `$currentUser instanceof PasswordAuthenticatedUserInterface`

File: src/Symfony/Component/Security/Core/Validator/Constraints/UserPasswordValidator.php [hunk 4]
```diff
@@ -53,22 +52,13 @@ public function validate(mixed $password, Constraint $constraint)
 
         $user = $this->tokenStorage->getToken()->getUser();
 
-        if (!$user instanceof UserInterface) {
-            throw new ConstraintDefinitionException('The User object must implement the UserInterface interface.');
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Validator/Constraints/UserPasswordValidator.php:57]: is it intended to remove this error ?

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 3 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: done

File: src/Symfony/Component/Security/Core/Validator/Constraints/UserPasswordValidator.php (hunk 4 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Validator/Constraints/UserPasswordValidator.php:57]: Yes, we only need `getPassword()` here so the check is replaced accordingly

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 2 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: fixed

File: src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php (hunk 1 above)
Comment (chalasr) [reply src/Symfony/Component/PasswordHasher/Hasher/UserPasswordHasher.php]: agreed, reverted

**Comment by chalasr** [comment]
`high-deps` build should be fixed when merging. PR ready

**Comment by nicolas-grekas** [comment]
> high-deps build should be fixed when merging. PR ready

Actually not: the `^` after the name of the component is a reminder that this is a flipped test: checkout 5.4 and use with patched 6.0 deps.

Merging as is would break 5.4, that's what this tells.

**Comment by chalasr** [comment]
Thanks, on it.

**Comment by chalasr** [comment]
Should be ok with #42001.

**Comment by chalasr** [comment]
Now green

File: src/Symfony/Bridge/Doctrine/Tests/Fixtures/User.php [hunk 5]
```diff
@@ -58,6 +54,10 @@ public function getUserIdentifier(): string
         return $this->name;
     }
 
+    public function getSalt(): ?string
```
Comment (Tobion) [inline src/Symfony/Bridge/Doctrine/Tests/Fixtures/User.php]: this method is still needed? why move it inside the class?

File: src/Symfony/Component/Security/Core/Tests/Authentication/Provider/DaoAuthenticationProviderTest.php [hunk 6]
```diff
@@ -338,7 +339,7 @@ protected function getProvider($user = null, $userChecker = null, $passwordHashe
     }
 }
 
-class TestUser implements UserInterface
+class TestUser implements PasswordAuthenticatedUserInterface, UserInterface
```
Comment (Tobion) [inline src/Symfony/Component/Security/Core/Tests/Authentication/Provider/DaoAuthenticationProviderTest.php:342]: function getSalt can be removed from this class or?

Comment (chalasr) [reply src/Symfony/Component/Security/Core/Tests/Authentication/Provider/DaoAuthenticationProviderTest.php:342]: yes indeed, done

File: src/Symfony/Bridge/Doctrine/Tests/Fixtures/User.php (hunk 5 above)
Comment (chalasr) [reply src/Symfony/Bridge/Doctrine/Tests/Fixtures/User.php]: Bad rebase, now removed from the class.

**Comment by chalasr** [comment]
Comments addressed, thanks for the reviews.

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php [hunk 7]
```diff
@@ -51,8 +51,13 @@ public function __construct(UserProviderInterface $userProvider, UserCheckerInte
      */
     protected function checkAuthentication(UserInterface $user, UsernamePasswordToken $token)
     {
+        if (!$user instanceof PasswordAuthenticatedUserInterface) {
+            throw new \TypeError(sprintf('Class "%s" must implement "%s" for using password-based authentication.', get_debug_type($user), PasswordAuthenticatedUserInterface::class));
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: I think this shouldn't be a `TypeError`, but `InvalidArgumentException` instead. Or at least, I'm using these rules myself:

* If it's really a bad type from the contract point of view (but I can't add the type to the argument), I'm using `TypeError`.
* If it's the correct type, but it's correct from the contract/interface point of view, I'm using `InvalidArgumentException`.

Not sure if it makes sense though, they seem very related

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php [hunk 8]
```diff
@@ -273,14 +274,16 @@ private function hasUserChanged(UserInterface $user): bool
...
+            }
 
-        // @deprecated since Symfony 5.3, check for LegacyPasswordAuthenticatedUserInterface on both user objects before comparing salts
-        if ($this->user->getSalt() !== $user->getSalt()) {
-            return true;
+            if ($this->user instanceof LegacyPasswordAuthenticatedUserInterface || $user instanceof LegacyPasswordAuthenticatedUserInterface) {
+                if (!$this->user instanceof LegacyPasswordAuthenticatedUserInterface || !$user instanceof LegacyPasswordAuthenticatedUserInterface || $this->user->getSalt() !== $user->getSalt()) {
+                    return true;
+                }
+            }
```
Comment (wouterj) [inline src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: What about?

```suggestion
            if ($this->user instanceof LegacyPasswordAuthenticatedUserInterface xor $user instanceof LegacyPasswordAuthenticatedUserInterface) {
                return true;
            }

            if ($this->user instanceof LegacyPasswordAuthenticatedUserInterface && $this->user->getSalt() !== $user->getSalt()) {
                return true;
            }
```

File: src/Symfony/Component/Security/Http/RememberMe/TokenBasedRememberMeServices.php [hunk 9]
```diff
@@ -61,6 +62,10 @@ protected function processAutoLoginCookie(array $cookieParts, Request $request)
             throw new \RuntimeException(sprintf('The UserProviderInterface implementation must return an instance of UserInterface, but returned "%s".', get_debug_type($user)));
         }
 
+        if (!$user instanceof PasswordAuthenticatedUserInterface) {
+            throw new \RuntimeException(sprintf('Class "%s" must implement "%s" for using "%s".', get_debug_type($user), PasswordAuthenticatedUserInterface::class, __CLASS__));
+        }
```
Comment (wouterj) [inline src/Symfony/Component/Security/Http/RememberMe/TokenBasedRememberMeServices.php:67]: I don't like this: remember shouldn't depend on passwords imho (e.g. when using login links or webauthn)

Comment (wouterj) [reply src/Symfony/Component/Security/Http/RememberMe/TokenBasedRememberMeServices.php:67]: Oh, nvm this class is going to be removed anyways (it's the legacy remember me service)

File: src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php (hunk 7 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Authentication/Provider/DaoAuthenticationProvider.php]: I agree it fits better

File: src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php (hunk 8 above)
Comment (chalasr) [reply src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php]: looks nice :) updated

**Review by wouterj — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @chalasr.

**Pull request merged**